src/storage/recovery/recovery_manager.cpp

```diff
@@ -0,0 +1,934 @@
+        delete[] reinterpret_cast<byte *>(log_record);
+        break;
+      }
+
+      default:
+        TERRIER_ASSERT(
+            log_record->RecordType() == LogRecordType::REDO || log_record->RecordType() == LogRecordType::DELETE,
+            "We should only buffer changes for redo or delete records");
+        buffered_changes_map_[log_record->TxnBegin()].push_back(pair);
+    }
+  }
+  // Process all deferred txns
+  ProcessDeferredTransactions(transaction::NO_ACTIVE_TXN);
+  TERRIER_ASSERT(deferred_txns_.empty(), "We should have no unprocessed deferred transactions at the end of recovery");
+
+  // If we have unprocessed buffered changes, then these transactions were in-process at the time of system shutdown.
+  // They are unrecoverable, so we need to clean up the memory of their records.
+  if (!buffered_changes_map_.empty()) {
+    for (auto &txn : buffered_changes_map_) {
+      for (auto &buffered_pair : txn.second) {
+        delete[] reinterpret_cast<byte *>(buffered_pair.first);
+        for (auto *entry : buffered_pair.second) {
+          delete[] entry;
+        }
+      }
+    }
+  }
+
+  return txns_replayed;
+}
+
+void RecoveryManager::ProcessCommittedTransaction(terrier::transaction::timestamp_t txn_id) {
+  // Begin a txn to replay changes with.
+  auto *txn = txn_manager_->BeginTransaction();
+
+  // Apply all buffered changes. They should all succeed. After applying we can safely delete the record
+  for (uint32_t idx = 0; idx < buffered_changes_map_[txn_id].size(); idx++) {
+    auto *buffered_record = buffered_changes_map_[txn_id][idx].first;
+    TERRIER_ASSERT(
+        buffered_record->RecordType() == LogRecordType::REDO || buffered_record->RecordType() == LogRecordType::DELETE,
+        "Buffered record must be a redo or delete.");
+
+    if (IsSpecialCaseCatalogRecord(buffered_record)) {
+      idx += ProcessSpecialCaseCatalogRecord(txn, &buffered_changes_map_[txn_id], idx);
+    } else if (buffered_record->RecordType() == LogRecordType::REDO) {
+      ReplayRedoRecord(txn, buffered_record);
+    } else {
+      ReplayDeleteRecord(txn, buffered_record);
+    }
+
+    delete[] reinterpret_cast<byte *>(buffered_record);
+  }
+  buffered_changes_map_.erase(txn_id);
+  // Commit the txn
+  txn_manager_->Commit(txn, transaction::TransactionUtil::EmptyCallback, nullptr);
+}
+
+void RecoveryManager::ProcessAbortedTransaction(terrier::transaction::timestamp_t txn_id) {
+  // If we are aborting, we can free and discard all buffered changes. Nothing needs to be replayed
+  for (auto &buffered_pair : buffered_changes_map_[txn_id]) {
+    delete[] reinterpret_cast<byte *>(buffered_pair.first);
+    for (auto *entry : buffered_pair.second) {
+      delete[] entry;
+    }
+  }
+  buffered_changes_map_.erase(txn_id);
+}
+
+void RecoveryManager::DeferTransaction(terrier::transaction::timestamp_t txn_id) {
+  // As an optimization, because we process txns in nearly monotonically increasing order, we iterate from the end to
```

Comment:
Is this a premature optimization?  If we need ordering, wouldn't it make more sense to use a sorted container?

Reply:
I didn't realize there were containers in STL. I'll use [`std::set`](https://en.cppreference.com/w/cpp/container/set).